{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "28dc675b",
   "metadata": {},
   "source": [
    "# Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "29d11757",
   "metadata": {},
   "outputs": [],
   "source": [
    "from helpers import SEABORN_COLORS, create_plot\n",
    "import os\n",
    "import sys\n",
    "import time\n",
    "import datetime\n",
    "\n",
    "from termcolor import colored, cprint\n",
    "\n",
    "os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'\n",
    "import sys\n",
    "from os import listdir\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd  # Import Pandas and Numpy to create databases\n",
    "import tensorflow as tf  # Import tensorflow to do tensor manipulations\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "import matplotlib.pyplot as plt\n",
    "import csv\n",
    "import statsmodels.api as sm\n",
    "import seaborn as sns\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dca3d19",
   "metadata": {},
   "source": [
    "# PATHS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "27552e00",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/mnt/d/UKB_toydata/usefull_values_2’: File exists\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "256"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PATH_data = f\"/mnt/d/UKB_toydata/\"\n",
    "PATH_GENO = f\"{PATH_data}genotype\"\n",
    "PATH_usefull = f\"{PATH_data}usefull_values\"\n",
    "PATH_output = f\"{PATH_data}usefull_values_2\"\n",
    "os.system(f\"mkdir {PATH_output}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d778b6a8",
   "metadata": {},
   "source": [
    "# Start loop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "95cfb31b",
   "metadata": {},
   "outputs": [],
   "source": [
    "diseases = listdir(PATH_usefull)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b878f0d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLINK v2.00a5LM 64-bit Intel (21 Jun 2023)     www.cog-genomics.org/plink/2.0/\n",
      "(C) 2005-2023 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to ./plink.log.\n",
      "Options in effect:\n",
      "  --bfile /mnt/d/UKB_toydata/genotype/02_per_chrom/total_pop_chr_22\n",
      "  --export A\n",
      "  --geno 0.05\n",
      "  --mac 50\n",
      "  --maf 0\n",
      "  --max-maf 0.01\n",
      "  --out ./plink\n",
      "\n",
      "Start time: Thu Sep 14 20:23:59 2023\n",
      "7857 MiB RAM detected, ~6213 available; reserving 3928 MiB for main workspace.\n",
      "Using up to 12 threads (change this with --threads).\n",
      "69455 samples (0 females, 0 males, 69455 ambiguous; 69455 founders) loaded from\n",
      "/mnt/d/UKB_toydata/genotype/02_per_chrom/total_pop_chr_22.fam.\n",
      "16775 variants loaded from\n",
      "/mnt/d/UKB_toydata/genotype/02_per_chrom/total_pop_chr_22.bim.\n",
      "Note: No phenotype data present.\n",
      "Calculating allele frequencies... done.\n",
      "--geno: 2 variants removed due to missing genotype data.\n",
      "16649 variants removed due to allele frequency threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "124 variants remaining after main filters.\n",
      "--export A pass 1/1: loading... 549writing... 10101111121213131414151516161717181819202021212222232324242525262627272828292930303131323233333434353536363737383839404041414242434344444545464647474848494950505151525253535454555556565757585859606061616262636364646565666667676868696970707171727273737474757576767777787879808081818282838384848585868687878888898990909191929293939494959596969797989899done.\n",
      "--export A: ./plink.raw written.\n",
      "End time: Thu Sep 14 20:24:00 2023\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/mnt/d/UKB_toydata/usefull_values_2/no_corr’: File exists\n",
      "mkdir: cannot create directory ‘/mnt/d/UKB_toydata/usefull_values_2/no_corr//bc’: File exists\n"
     ]
    }
   ],
   "source": [
    "PATH_input = PATH_GENO+ \"/02_per_chrom/\"\n",
    "name = \"total_pop\"\n",
    "chrom = 22\n",
    "plink_snps_pool = 20\n",
    "os.system(f\"plink2 --bfile {PATH_input}{name}_chr_{chrom} --maf 0 --max-maf 0.01 --mac 50 --geno 0.05 --recode A --out ./plink\")\n",
    "id_samples = pd.read_csv('./plink.raw', sep=\"\\t\")[['FID']]\n",
    "os.system(\"rm plink.raw\")\n",
    "\n",
    "\n",
    "# NO CORRECTION\n",
    "os.system(f\"mkdir {PATH_output}/no_corr\")\n",
    "temp_path = f\"{PATH_output}/no_corr/\"\n",
    "\n",
    "\n",
    "\n",
    "chroms = [21,22]\n",
    "for disease in diseases:\n",
    "    PATH_pheno = f\"/mnt/d/UKB_toydata/\"\n",
    "    PATH_pheno = PATH_pheno + f\"old/{disease}_eur_sample0.5.fam\"\n",
    "    columns = ['IID','FID','na1','na2','na3','phenotype']\n",
    "    pheno_with_id = pd.read_csv(PATH_pheno, sep=\"\\t\", header=None, names=columns)\n",
    "    #pheno_with_id = pd.read_csv(PATH_pheno, sep=\"\\t\", header=None, names=columns)[['FID','phenotype']]\n",
    "    pheno = id_samples.merge(pheno_with_id, how='inner', on='FID')\n",
    "    \n",
    "    PRS = np.array([0] * pheno.shape[0])\n",
    "    snps = []\n",
    "    coefs = []\n",
    "    std_errs = []\n",
    "    ts = []\n",
    "    Ps = []\n",
    "    twfives = []\n",
    "    ninfives = []\n",
    "    os.system(f\"mkdir {temp_path}/{disease}\")\n",
    "    dims = 100\n",
    "    for chrom in chroms:\n",
    "        path_genos = PATH_GENO\n",
    "        path_genos = path_genos + \"/04_LD_blocks/\" + f\"{dims}_PCs/chrom_{chrom}/\"\n",
    "        for file in listdir(path_genos):\n",
    "            path_geno = path_genos + file\n",
    "            genos = pd.read_pickle(path_geno)\n",
    "            genos['FID'] = id_samples['FID']\n",
    "            genos = genos.merge(pheno_with_id[['FID']], how='inner', on='FID')\n",
    "            genos = genos.drop(columns='FID')\n",
    "            for snp_id in list(genos.columns):\n",
    "                #print('-------------------')\n",
    "                #print(snp_id)\n",
    "                #print(genos[snp_id])\n",
    "                X2 = sm.add_constant(genos[[snp_id]])\n",
    "                est = sm.OLS(pheno[['phenotype']], X2)\n",
    "                est2 = est.fit()\n",
    "                snps.append(est2.pvalues.index[1])\n",
    "                coefs.append(est2.params[1])\n",
    "                std_errs.append(est2.bse[1])\n",
    "                ts.append(est2.tvalues[1])\n",
    "\n",
    "                P = est2.pvalues[1]\n",
    "                logp = - np.log(P)\n",
    "                if logp > 0:\n",
    "                    PRS = PRS + np.array(est2.tvalues[1]*genos[snp_id])\n",
    "                else:\n",
    "                    pass\n",
    "                Ps.append(P)\n",
    "                twfives.append(est2.conf_int()[0][1])\n",
    "                ninfives.append(est2.conf_int()[1][1])\n",
    "    sumstat = pd.DataFrame(data={\n",
    "                'snp': snps,\n",
    "                'coef': coefs,\n",
    "                'std_err': std_errs,\n",
    "                'Z': ts,\n",
    "                'P': Ps,\n",
    "                '0.025': twfives,\n",
    "                '0.975': ninfives\n",
    "            })\n",
    "    sumstat['#CHROM'] = sumstat['snp'].str[:2].astype(int)\n",
    "    sumstat['POS'] = [i.split(':', 1)[0] for i in list(sumstat['snp'].str[3:-5])]\n",
    "    sumstat['POS'] = sumstat['POS'].astype(int)\n",
    "    sumstat['-logp'] = - np.log(sumstat['P'])\n",
    "    \n",
    "    sumstat['#CHROM'] = [f.split(':')[0] for f in list(sumstat['snp'])]\n",
    "    sumstat['POS'] = [f.split(':')[1] for f in list(sumstat['snp'])]\n",
    "    sumstat['POS'] = sumstat['POS'].astype(int)\n",
    "    running_pos = 0\n",
    "    cumulative_pos = []\n",
    "    for chrom, group_df in sumstat.groupby('#CHROM'):\n",
    "        cumulative_pos.append(group_df['POS'] + running_pos)\n",
    "        running_pos += group_df['POS'].max()\n",
    "\n",
    "    sumstat['cumulative_pos'] = pd.concat(cumulative_pos)\n",
    "    sumstat['SNP number'] = sumstat.index\n",
    "    pheno['PRS'] = PRS\n",
    "    sumstat.to_pickle(f\"{temp_path}/{disease}/sumstat.pkl\")\n",
    "    pheno.to_pickle(f\"{temp_path}/{disease}/pheno_PRS.pkl\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83dac3a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Abyss CORRECTION\n",
    "os.system(f\"mkdir {PATH_output}/no_corr\")\n",
    "temp_path = f\"{PATH_output}/no_corr/\"\n",
    "\n",
    "PRS = np.array([0] * pheno.shape[0])\n",
    "snps = []\n",
    "coefs = []\n",
    "std_errs = []\n",
    "ts = []\n",
    "Ps = []\n",
    "twfives = []\n",
    "ninfives = []\n",
    "dims = [1, 2, 7, 15, 25, 100]\n",
    "for dim in dims:\n",
    "    chroms = [21,22]\n",
    "    for disease in diseases[-2:-1]:\n",
    "        PATH_pheno = f\"/mnt/d/UKB_toydata/\"\n",
    "        PATH_pheno = PATH_pheno + f\"old/{disease}_eur_sample0.5.fam\"\n",
    "        columns = ['IID','FID','na1','na2','na3','phenotype']\n",
    "        pheno_with_id = pd.read_csv(PATH_pheno, sep=\"\\t\", header=None, names=columns)\n",
    "        #pheno_with_id = pd.read_csv(PATH_pheno, sep=\"\\t\", header=None, names=columns)[['FID','phenotype']]\n",
    "        pheno = id_samples.merge(pheno_with_id, how='inner', on='FID')\n",
    "        os.system(f\"mkdir {temp_path}/{disease}\")\n",
    "        dims = 100\n",
    "        for chrom in chroms:\n",
    "            path_genos = PATH_GENO\n",
    "            path_genos = path_genos + \"/04_LD_blocks/\" + f\"{dim}_PCs/chrom_{chrom}/\"\n",
    "            for file in listdir(path_genos)[0:2]:\n",
    "                path_geno = path_genos + file\n",
    "                genos = pd.read_pickle(path_geno)\n",
    "                genos['FID'] = id_samples['FID']\n",
    "                genos = genos.merge(pheno_with_id[['FID']], how='inner', on='FID')\n",
    "                genos = genos.drop(columns='FID')\n",
    "                for snp_id in list(genos.columns)[0:100]:\n",
    "                    #print('-------------------')\n",
    "                    #print(snp_id)\n",
    "                    #print(genos[snp_id])\n",
    "                    X2 = sm.add_constant(genos[[snp_id]])\n",
    "                    est = sm.OLS(pheno[['phenotype']], X2)\n",
    "                    est2 = est.fit()\n",
    "                    snps.append(est2.pvalues.index[1])\n",
    "                    coefs.append(est2.params[1])\n",
    "                    std_errs.append(est2.bse[1])\n",
    "                    ts.append(est2.tvalues[1])\n",
    "\n",
    "                    P = est2.pvalues[1]\n",
    "                    logp = - np.log(P)\n",
    "                    if logp > 0:\n",
    "                        PRS = PRS + np.array(est2.tvalues[1]*genos[snp_id])\n",
    "                    else:\n",
    "                        pass\n",
    "                    Ps.append(P)\n",
    "                    twfives.append(est2.conf_int()[0][1])\n",
    "                    ninfives.append(est2.conf_int()[1][1])\n",
    "        sumstat = pd.DataFrame(data={\n",
    "                    'snp': snps,\n",
    "                    'coef': coefs,\n",
    "                    'std_err': std_errs,\n",
    "                    'Z': ts,\n",
    "                    'P': Ps,\n",
    "                    '0.025': twfives,\n",
    "                    '0.975': ninfives\n",
    "                })\n",
    "        sumstat['#CHROM'] = sumstat['snp'].str[:2].astype(int)\n",
    "        sumstat['POS'] = [i.split(':', 1)[0] for i in list(sumstat['snp'].str[3:-5])]\n",
    "        sumstat['POS'] = sumstat['POS'].astype(int)\n",
    "        sumstat['-logp'] = - np.log(sumstat['P'])\n",
    "\n",
    "        sumstat['#CHROM'] = [f.split(':')[0] for f in list(sumstat['snp'])]\n",
    "        sumstat['POS'] = [f.split(':')[1] for f in list(sumstat['snp'])]\n",
    "        sumstat['POS'] = sumstat['POS'].astype(int)\n",
    "        running_pos = 0\n",
    "        cumulative_pos = []\n",
    "        for chrom, group_df in sumstat.groupby('#CHROM'):\n",
    "            cumulative_pos.append(group_df['POS'] + running_pos)\n",
    "            running_pos += group_df['POS'].max()\n",
    "\n",
    "        sumstat['cumulative_pos'] = pd.concat(cumulative_pos)\n",
    "        sumstat['SNP number'] = sumstat.index\n",
    "        pheno['PRS'] = PRS\n",
    "        sumstat.to_pickle(f\"{temp_path}/{disease}/sumstat.pkl\")\n",
    "        pheno.to_pickle(f\"{temp_path}/{disease}/pheno_PRS.pkl\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
